examples/hosts/app-integration/schema-upgrade/src/migrator/migrator.ts

```diff
@@ -86,100 +108,133 @@ export class Migrator extends TypedEventEmitter<IMigratorEvents> implements IMig
-            // above await.
-            if (migratable.migrationTool.migrationState as MigrationState === "migrated") {
-                this._migrationP = undefined;
-                migratedModel.close();
-                this.takeAppropriateActionForCurrentMigratable();
-                return;
-            }
+            const completeTheMigration = async () => {
+                assert(this._preparedDetachedModel !== undefined, "this._preparedDetachedModel should be defined");
 
-            // TODO: Support retry
-            await migratable.migrationTool.finalizeMigration(containerId);
+                // Volunteer to complete the migration.
+                let isAssigned: boolean;
+                try {
+                    isAssigned = await this.currentModel.migrationTool.volunteerForMigration();
+                } catch (error) {
+                    // volunteerForMigration() will throw an error on disconnection. In this case, we should exit and
+                    // re-enter the state machine which will wait until we reconnect.
+                    // Note: while we wait to reconnect it is possible that another client will have already completed
+                    // the migration.
+                    assert(!this.connected, "We should be disconnected");
+                    return;
+                }
 
-            // If someone else finalized the migration before us, we should close the one we created.
-            if (migratable.migrationTool.newContainerId !== containerId) {
-                migratedModel.close();
-            }
+                if (!isAssigned) {
+                    // If we are not assigned, then the migration was completed by another client and we can exit.
+                    return;
+                }
 
-            // Note that we do not assume the migratedModel is the correct new one, and let it fall out of scope
-            // intentionally.  This is because if we don't win the race to set the container, it will be the wrong
-            // container/model to use.  There could maybe be some efficiency gain by retaining the model in the
-            // case that we win the race?  But it likely just doesn't matter that much because the Loader probably
-            // cached the Container anyway.
+                if (this._containerId === undefined) {
+                    this._containerId = await this._preparedDetachedModel.attach();
+                }
 
-            this._migrationP = undefined;
+                // Check to make sure we still have the task assignment.
+                if (!this.currentModel.migrationTool.haveMigrationTask()) {
+                    // Exit early if we lost the task assignment, we are most likely disconnected.
+                    return;
+                }
 
-            this.takeAppropriateActionForCurrentMigratable();
+                await migratable.migrationTool.finalizeMigration(this._containerId);
```

Comment:
What would play out here if you lose the lock?

Reply:
We would end up throwing an error on the next line (`completeMigraitonTask()`), since we are not able to complete a task without being assigned. I think we should probably try to prevent that with another check though

Reply:
As suggested above, would promise race be a general pattern that helps us avoid these checks?